File: apps/api/app/routers/dataset.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.middleware.auth import Actor, require_roles
from app.services import dataset_domain
from app.schemas.dataset import (
    ImportDatasetRequest,
    ImportDatasetResponse,
    DatasetItemResponse,
    ListItemsResponse,
    UpdateDatasetItemRequest,
)

router = APIRouter(tags=["dataset"])


@router.post(
    "/tasks/{task_id}/dataset/import",
    response_model=ImportDatasetResponse,
    status_code=201,
    summary="导入数据集（JSON / JSONL / Excel）",
)
def import_dataset(
    task_id: str,
    body: ImportDatasetRequest,
    db: Session = Depends(get_db),
    actor: Actor = Depends(require_roles("OWNER")),
) -> ImportDatasetResponse:
    """从已上传文件导入数据集（契约 §23.2）。文件必须 purpose=DATASET_IMPORT、status=READY。"""
    return dataset_domain.import_dataset(db, task_id, actor, body)


@router.get(
    "/tasks/{task_id}/items",
    response_model=ListItemsResponse,
    summary="获取题目列表（分页）",
)
def list_items(
    task_id: str,
    page: int = Query(1, ge=1, description="页码，从 1 开始"),
    pageSize: int = Query(20, ge=1, le=200, description="每页数量"),
    db: Session = Depends(get_db),
    actor: Actor = Depends(require_roles("OWNER")),
) -> ListItemsResponse:
    return dataset_domain.list_items(db, task_id, actor, page, pageSize)


@router.get(
    "/items/{item_id}",
    response_model=DatasetItemResponse,
    summary="获取题目详情",
)
def get_item(
    item_id: str,
    db: Session = Depends(get_db),
    actor: Actor = Depends(require_roles("OWNER", "REVIEWER", "LABELER")),
) -> DatasetItemResponse:
    """OWNER / REVIEWER / LABELER 均可访问（契约 §23.2）。"""
    item = dataset_domain.get_item(db, item_id, actor)
    return DatasetItemResponse.from_orm(item)


@router.patch(
    "/items/{item_id}",
    response_model=DatasetItemResponse,
    summary="编辑题目",
)
def update_item(
    item_id: str,
    body: UpdateDatasetItemRequest,
    db: Session = Depends(get_db),
    actor: Actor = Depends(require_roles("OWNER")),
) -> DatasetItemResponse:
    """更新题目 sourcePayload 或 status（status 仅限 AVAILABLE/DISABLED）。"""
    item = dataset_domain.update_item(db, item_id, actor, body)
    return DatasetItemResponse.from_orm(item)

File: apps/web/src/api/dataset.ts
import type { DatasetItem } from "@labelhub/contracts";
import { apiGet, apiPost, apiPatch, apiUploadBinary } from "./client";

export type DatasetFormat = "JSON" | "JSONL" | "EXCEL";

export interface ImportDatasetResult {
  taskId: string;
  importedCount: number;
  skippedCount: number;
  failedCount: number;
  previewItems: DatasetItem[];
  errors?: { row?: number; message: string }[];
}

export interface ListItemsResult {
  items: DatasetItem[];
  total: number;
  page: number;
  pageSize: number;
}

function currentActorId(): string {
  try {
    const raw = localStorage.getItem("labelhub_actor");
    if (raw) return (JSON.parse(raw) as { id?: string }).id ?? "";
  } catch {
    /* ignore */
  }
  return "";
}

/** 依扩展名推断导入格式 */
export function inferFormat(fileName: string): DatasetFormat {
  const lower = fileName.toLowerCase();
  if (lower.endsWith(".jsonl")) return "JSONL";
  if (lower.endsWith(".xlsx") || lower.endsWith(".xls")) return "EXCEL";
  return "JSON";
}

/**
 * 完整数据集导入流程（对接已就绪后端）：
 * 1) POST /files/upload-url 申请上传  2) 上传二进制  3) /confirm 置 READY
 * 4) POST /tasks/{id}/dataset/import 解析入库
 */
export async function importDataset(
  taskId: string,
  file: File,
  format: DatasetFormat,
  externalKeyPath?: string,
): Promise<ImportDatasetResult> {
  const contentType = file.type || "application/octet-stream";
  const created = await apiPost<{ file: { id: string }; uploadUrl: string }>(
    "/api/v1/files/upload-url",
    {
      fileName: file.name,
      mimeType: contentType,
      size: file.size,
      purpose: "DATASET_IMPORT",
      ownerType: "USER",
      ownerId: currentActorId(),
    },
  );
  const fileId = created.file.id;
  // 走相对路径经 Vite 代理上传，避免后端返回的绝对 uploadUrl 在容器/宿主间不一致
  await apiUploadBinary(`/api/v1/files/${fileId}/upload`, file, contentType);
  await apiPost(`/api/v1/files/${fileId}/confirm`, {});

  const body: Record<string, unknown> = { fileId, format };
  if (externalKeyPath) body.externalKeyPath = externalKeyPath;
  return apiPost<ImportDatasetResult>(`/api/v1/tasks/${taskId}/dataset/import`, body);
}

export async function listItems(taskId: string, page = 1, pageSize = 50): Promise<ListItemsResult> {
  return apiGet<ListItemsResult>(`/api/v1/tasks/${taskId}/items?page=${page}&pageSize=${pageSize}`);
}

export async function updateItem(
  itemId: string,
  patch: { sourcePayload?: Record<string, unknown>; status?: "AVAILABLE" | "DISABLED" },
): Promise<DatasetItem> {
  return apiPatch<DatasetItem>(`/api/v1/items/${itemId}`, patch);
}
